pkg/resourcemanager/rediscaches/rediscacheaction_reconcile.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package rediscaches
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-service-operator/api/v1alpha1"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (m *AzureRedisCacheActionManager) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	options := &resourcemanager.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	instance, err := m.convert(obj)
+	if err != nil {
+		return true, err
+	}
+
+	// never re-provision an action
+	if instance.Status.Provisioned {
+		return true, nil
+	}
+
+	rollAllKeys := instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollAllKeys
+
+	if rollAllKeys || instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollPrimaryKey {
+		if err = m.RegeneratePrimaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
+			return false, err
+		}
+	}
+
+	if rollAllKeys || instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollSecondaryKey {
+		if err = m.RegenerateSecondaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
+			return false, err
+		}
+	}
+
+	// regenerate the secret
+	cacheInstance := &v1alpha1.RedisCache{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Spec.CacheName,
+			Namespace: instance.Namespace,
+		},
+		Spec: v1alpha1.RedisCacheSpec{
+			SecretName:        instance.Spec.SecretName,
+			ResourceGroupName: instance.Spec.ResourceGroup,
+		},
+	}
+	if err = m.ListKeysAndCreateSecrets(ctx, cacheInstance); err != nil {
+		instance.Status.Provisioned = false
+		instance.Status.FailedProvisioning = true
+		return false, err
+	}
+
+	// successful return
+	instance.Status.Provisioned = true
+	instance.Status.FailedProvisioning = false
+	return true, nil
+}
+
+func (m *AzureRedisCacheActionManager) Delete(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	// no deletion necessary for deletion of action
+	return false, nil
+}
+
+func (m *AzureRedisCacheActionManager) GetParents(obj runtime.Object) ([]resourcemanager.KubeParent, error) {
+	instance, err := m.convert(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return []resourcemanager.KubeParent{
+		{
+			Key: types.NamespacedName{
+				Namespace: instance.Namespace,
+				Name:      instance.Spec.ResourceGroup,
+			},
+			Target: &v1alpha1.ResourceGroup{},
+		},
+		{
+			Key: types.NamespacedName{
+				Namespace: instance.Namespace,
+				Name:      instance.Spec.CacheName,
+			},
+			Target: &v1alpha1.RedisCache{},
+		},
```

Comment:
I think your order here is in reverse. The Async Controller loops through the parents until it finds one that it can reference as owner. In this case it would reference the RG before trying the RedisCache itself.

Reply:
I'm still getting the RedisCache as the owner when I deploy locally. Would it be better to only declare the rediscache as a parent since it doesn't really have a resource group?

Reply:
Having all the possible ancestors allows for the possibility that the users might choose to create some of the dependencies out of band. I don't know why they might do this but they might create the RG, then create the redis some other way for somme reason, then maybe they still want to try to run RedisActions. In this case the Action would be owned by the RG and would be cleaned up if the RG was deleted from kube.

Reply:
Fixed.